pkg/cassandra/config.go

```diff
@@ -3,153 +3,99 @@ package cassandra
 import (
 	"encoding/json"
 	"fmt"
+	"github.com/Masterminds/semver/v3"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/unstructured"
 	"reflect"
 	"sort"
-	"strconv"
 	"strings"
 
-	"k8s.io/utils/pointer"
-
 	api "github.com/k8ssandra/k8ssandra-operator/apis/k8ssandra/v1alpha1"
-	"github.com/k8ssandra/k8ssandra-operator/pkg/encryption"
 )
 
 const (
 	SystemReplicationFactorStrategy = "-Dcassandra.system_distributed_replication"
 	allowAlterRf                    = "-Dcassandra.allow_alter_rf_during_range_movement=true"
 )
 
-// CreateJsonConfig parses a DatacenterConfig into a raw JSON base64-encoded string as required by
-// the CassandraDatacenter.Spec.Config field, which is processed by cass-config-builder.
-func CreateJsonConfig(apiConfig *DatacenterConfig) ([]byte, error) {
-	cfg := &cassConfig{
-		CassandraConfig: apiConfig.CassandraConfig.DeepCopy(),
+// createJsonConfig parses a CassandraConfig into raw JSON bytes as required by the
+// CassandraDatacenter.Spec.Config field, which is processed by cass-config-builder.
+func createJsonConfig(config api.CassandraConfig, serverVersion *semver.Version, serverType api.ServerDistribution) ([]byte, error) {
+
+	out := make(unstructured.Unstructured)
+
+	// cassandra.yaml is an unstructured map, we simply append it to the output
+	// TODO validate cassandra.yaml settings with JSON schemas
+	// TODO postprocess cassandra.yaml settings, e.g. mountable volumes or resource.Quantity syntax
+	if len(config.CassandraYaml) > 0 {
+		out["cassandra-yaml"] = config.CassandraYaml
 	}
-	addNumTokens(apiConfig, cfg)
```

Comment:
I moved all mutating functions outside of this function; I would like this function to be devoted solely to producing the final JSON by marshaling the input config.